### PR TITLE
Implement delete operation in BSTree

### PR DESCRIPTION
**Pull request by ksenmel**
Implemented ``delete`` method in Simple BSTree. 
No automatic testing was done.

### CODE REVIEW

File: lib/src/main/kotlin/bstrees/SimpleBST.kt [hunk 1]
```diff
@@ -5,7 +5,68 @@ import bstrees.nodes.SimpleNode
 class SimpleBST<T : Comparable<T>> : BinarySearchTree<T, SimpleNode<T>>() {
     override fun createNewNode(data: T) = SimpleNode(data)
 
+    /**
+     * Deletes node and returns it as a result (or null).
+     */
```
Comment (aartdem) [inline lib/src/main/kotlin/bstrees/SimpleBST.kt]: Please write documentation in one line like other functions

File: lib/src/main/kotlin/bstrees/SimpleBST.kt [hunk 2]
```diff
@@ -5,7 +5,68 @@ import bstrees.nodes.SimpleNode
...
+            val nodeToReplaceWith = if (node.left == null) node.right!! else node.left!!
+            if (node == parent.right) parent.right = nodeToReplaceWith
+            else parent.left = nodeToReplaceWith
+
+            nodeToReplaceWith.parent = parent
+        }
+    }
+
+    /** Searches for node's successor until the node value is placed on the leaf of the tree. */
+    private fun findMinimumNode(node: SimpleNode<T>): SimpleNode<T> {
```
Comment (aartdem) [inline lib/src/main/kotlin/bstrees/SimpleBST.kt]: Maybe findNodeToReplaceWith ? or something like that

File: lib/src/main/kotlin/bstrees/SimpleBST.kt [hunk 3]
```diff
@@ -5,7 +5,68 @@ import bstrees.nodes.SimpleNode
...
+
+    /** The node to be deleted has two children. */
+    private fun deleteNodeWithTwoChildren(node: SimpleNode<T>) {
+        var nodeToDelete = node
+        val nodeToReplaceWith = findMinimumNode(node.left!!)
+
+        nodeToDelete.data = nodeToReplaceWith.data
+        nodeToDelete = nodeToReplaceWith
+
+        deleteLeafNode(nodeToReplaceWith)
```
Comment (aartdem) [inline lib/src/main/kotlin/bstrees/SimpleBST.kt]: nodeToReplaceWith can be the node with one child. This case is not considered

**Review by aartdem — CHANGES_REQUESTED** [review]

File: lib/src/main/kotlin/bstrees/SimpleBST.kt [hunk 4]
```diff
@@ -5,7 +5,67 @@ import bstrees.nodes.SimpleNode
...
+                nodeToDelete.left == null && nodeToDelete.right == null -> deleteLeafNode(nodeToDelete)
+                nodeToDelete.left == null || nodeToDelete.right == null -> deleteNodeWithOneChild(nodeToDelete)
+                else -> deleteNodeWithTwoChildren(nodeToDelete)
+            }
+        }
+        return data
+    }
+
+    /** Searches for node to delete. Returns null if node is not found */
+    private fun searchNodeForDelete(data: T): SimpleNode<T>? = searchNode(data)
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/SimpleBST.kt]: What's the point of the function? It's the same as `searchNode`

File: lib/src/main/kotlin/bstrees/SimpleBST.kt [hunk 5]
```diff
@@ -5,7 +5,67 @@ import bstrees.nodes.SimpleNode
 class SimpleBST<T : Comparable<T>> : BinarySearchTree<T, SimpleNode<T>>() {
     override fun createNewNode(data: T) = SimpleNode(data)
 
+    /** Deletes node and returns it as a result (or null). */
     override fun delete(data: T): T? {
-        TODO("Not yet implemented")
+        val nodeToDelete = searchNodeForDelete(data)
+
+        if (nodeToDelete == null) return null
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/SimpleBST.kt]:
```suggestion
        val nodeToDelete = searchNode(data) ?: return null
```

File: lib/src/main/kotlin/bstrees/SimpleBST.kt [hunk 6]
```diff
@@ -5,7 +5,67 @@ import bstrees.nodes.SimpleNode
...
+        else {
+            when {
+                nodeToDelete.left == null && nodeToDelete.right == null -> deleteLeafNode(nodeToDelete)
+                nodeToDelete.left == null || nodeToDelete.right == null -> deleteNodeWithOneChild(nodeToDelete)
+                else -> deleteNodeWithTwoChildren(nodeToDelete)
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/SimpleBST.kt]:
```suggestion
                nodeToDelete.left == null && nodeToDelete.right == null ->
                    deleteLeafNode(nodeToDelete)
                    
                nodeToDelete.left == null || nodeToDelete.right == null ->
                    deleteNodeWithOneChild(nodeToDelete)
                    
                else -> deleteNodeWithTwoChildren(nodeToDelete)
```

File: lib/src/main/kotlin/bstrees/SimpleBST.kt [hunk 7]
```diff
@@ -5,7 +5,67 @@ import bstrees.nodes.SimpleNode
...
+        var nodeToReplaceWith = node
+        while (nodeToReplaceWith.right != null) {
+            nodeToReplaceWith = nodeToReplaceWith.right!!;
+        }
+        return nodeToReplaceWith;
+    }
+
+    /** The node to be deleted has two children. */
+    private fun deleteNodeWithTwoChildren(node: SimpleNode<T>) {
+        var nodeToDelete = node
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/SimpleBST.kt]: The variable is useless here

File: lib/src/main/kotlin/bstrees/SimpleBST.kt [hunk 8]
```diff
@@ -5,7 +5,67 @@ import bstrees.nodes.SimpleNode
...
+        val nodeToReplaceWith = findNodeToReplaceWith(node.left!!)
+
+        nodeToDelete.data = nodeToReplaceWith.data
+        nodeToDelete = nodeToReplaceWith
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/SimpleBST.kt]: This assignment doesn't do anything

File: lib/src/main/kotlin/bstrees/SimpleBST.kt [hunk 9]
```diff
@@ -5,7 +5,67 @@ import bstrees.nodes.SimpleNode
...
+        while (nodeToReplaceWith.right != null) {
+            nodeToReplaceWith = nodeToReplaceWith.right!!;
+        }
+        return nodeToReplaceWith;
+    }
+
+    /** The node to be deleted has two children. */
+    private fun deleteNodeWithTwoChildren(node: SimpleNode<T>) {
+        var nodeToDelete = node
+        val nodeToReplaceWith = findNodeToReplaceWith(node.left!!)
```
Comment (therain7) [inline lib/src/main/kotlin/bstrees/SimpleBST.kt]: I suggest you edit the function to accept the node to find replacement for.
```suggestion
        val nodeToReplaceWith = findNodeToReplaceWith(node)
```

**Review by therain7 — CHANGES_REQUESTED** [review]
Returning `data` in `delete` is incorrect behavior

**Review by aartdem — APPROVED** [review]

**Review by therain7 — APPROVED** [review]

**Pull request merged**